flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/ZeroCopyNettyMessageDecoder.java

```diff
@@ -0,0 +1,280 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.netty;
+
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.shaded.netty4.io.netty.buffer.ByteBuf;
+import org.apache.flink.shaded.netty4.io.netty.channel.ChannelHandlerContext;
+import org.apache.flink.shaded.netty4.io.netty.channel.ChannelInboundHandlerAdapter;
+
+import static org.apache.flink.runtime.io.network.netty.NettyMessage.FRAME_HEADER_LENGTH;
+import static org.apache.flink.runtime.io.network.netty.NettyMessage.MAGIC_NUMBER;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * Decodes messages from the fragmentary netty buffers. This decoder assumes the
+ * messages have the following format:
+ * +-----------------------------------+--------------------------------+
+ * | FRAME_HEADER ||  MESSAGE_HEADER   |     DATA BUFFER (Optional)     |
+ * +-----------------------------------+--------------------------------+
+ * and it decodes each part in order.
+ *
+ * This decoder tries best to eliminate copying. For the frame header and message header,
+ * it only cumulates data when they span multiple input buffers. For the buffer part, it
+ * copies directly to the input channels to avoid future copying.
+ *
+ * The format of the frame header is
+ * +------------------+------------------+--------+
+ * | FRAME LENGTH (4) | MAGIC NUMBER (4) | ID (1) |
+ * +------------------+------------------+--------+
+ */
+public class ZeroCopyNettyMessageDecoder extends ChannelInboundHandlerAdapter {
+
+    private static final int INITIAL_MESSAGE_HEADER_BUFFER_LENGTH = 128;
+
+    /** The parser to parse the message header. */
+    private final NettyMessageParser messageParser;
+
+    /** The buffer used to cumulate the frame header part. */
+    private ByteBuf frameHeaderBuffer;
+
+    /** The buffer used to receive the message header part. */
+    private ByteBuf messageHeaderBuffer;
+
+    /** Which part of the current message is being decoded. */
+    private DecodeStep decodeStep;
+
+    /** How many bytes have been decoded in current step. */
+    private int decodedBytesOfCurrentStep;
+
+    /** The intermediate state when decoding the current message. */
+    private final MessageDecodeIntermediateState intermediateState;
+
+    ZeroCopyNettyMessageDecoder(NettyMessageParser messageParser) {
+        this.messageParser = messageParser;
+        this.intermediateState = new MessageDecodeIntermediateState();
+    }
+
+    @Override
+    public void channelActive(ChannelHandlerContext ctx) throws Exception {
+        super.channelActive(ctx);
+
+        frameHeaderBuffer = ctx.alloc().directBuffer(NettyMessage.FRAME_HEADER_LENGTH);
```

Comment:
I thought of two things here:

- Considering the sender side while writing the header of `BufferResponse`, it would always allocate the buffer from the netty allocator for every message. On receiver side now we reuse the same header buffer every time. I am not sure whether we can get benefits from this change. If so, the sender side can also take this way future. If not, we do not need to explicitly maintain these two headers in decoder and it would make things more simple.

- If we want to maintain the static header buffers for every decoder, another option is to create direct memory directly, I mean bypassing the netty allocator. Because we ever wondered the memory overhead boost in netty internal management.

Another rough thought is if we can make all the memory allocation via the unified allocator, then it is easy for tracing all the memory usages. I mean also allocating these header buffers via existing `NetworkBufferAllocator` if possible.

Reply:
I think the writer side is different from the reader side in that we only have a message to deal with in the receiver side, however, we will have multiple sending messages at one time, and they all occupy header spaces. 

Beside, I think the reason to rely on the Netty `ByteBuf` is that to deal with different messages the `messageHeaderBuffer` needs to modify the capacity to be able to hold headers for different message. If we use a unPool buffer, we have to reimplement the logic to copy to a larger buffer. Besides, from Flink's perspective the Netty memory is allocated by the unit of Arena (Which is modified to 4M by us). This means as long as we do not consume more memory than 4M in Netty, it should be better to reuse the allocated memory.

Reply:
> I think the writer side is different from the reader side in that we only have a message to deal with in the receiver side, however, we will have multiple sending messages at one time, and they all occupy header spaces.

On write side, there is still only one header used at the same time. Once write&flush message by netty thread, if the socket is not writable that means the header message would not be recycled after writing, then the next message from partition would not continue writing into socket via judging `channel.isWritable()`. Once the channel becomes writable again to recycle the previous header buffer, then the next message would be written again via `PartitionRequestQueue#channelWritabilityChanged`. So it is the same case for both sender and receiver sides. Only one header buffer is actually used for every channel. Of course we can create a separate ticket for tracing the issue of reusing the same header buffer on sender side.

> Beside, I think the reason to rely on the Netty ByteBuf is that to deal with different messages the messageHeaderBuffer needs to modify the capacity to be able to hold headers for different message. If we use a unPool buffer, we have to reimplement the logic to copy to a larger buffer. 

I am not quite sure how much benefits we can get from the buffer resize implementation in netty stack. The header length size is fixed for `BufferResponse` and only `ErrorResponse` has unfixed length. But error message is less frequent and might not need reuse the same buffer always. We should consider the pros & cons for different options. From my point it is better to not rely on some internal implementation details of netty stack unless necessary. It is unsafe to expose the netty managed `ByteBuf` to outside which might cause inconsistent/leak issue if careless.

> Besides, from Flink's perspective the Netty memory is allocated by the unit of Arena (Which is modified to 4M by us). This causes that as long as we do not consume more memory than 4M in Netty, it should be better to reuse the allocated memory.

The chunk size adjustment should be orthogonal to this PR. Both ways would reduce the whole netty memory overhead in different cases. So it is better to measure them separately to see the effects. I would create a separate ticket for tracing this issue.